mne/io/tests/test_raw.py

```diff
@@ -1,14 +1,103 @@
 # Generic tests that all raw classes should run
 from os import path as op
-from numpy.testing import assert_allclose
+import math
+import numpy as np
+from numpy.testing import assert_allclose, assert_array_almost_equal
 
+from nose.tools import assert_equal, assert_true
+
+from mne import concatenate_raws
 from mne.datasets import testing
 from mne.io import Raw
+from mne.utils import _TempDir
+
+
+def _test_raw_reader(reader, test_preloading, test_blocks=False, **kwargs):
+    """Test reading, writing and slicing of raw classes.
+
+    Parameters
+    ----------
+    reader : function
+        Function to test.
+    test_preloading : bool
+        Whether not preloading is implemented for the reader. If True, both
+        cases and memory mapping to file are tested.
+    test_blocks : bool
+        Whether the data is divided into blocks with ``buffer_size_sec``
+        in the ``meas_info``.
+    **kwargs :
+        Arguments for the reader.
+
+    Returns
+    -------
+    raw : Instance of Raw
+        A preloaded Raw object.
+    """
+    tempdir = _TempDir()
+    raws = list()
+    raws.append(reader(**kwargs))
+    rng = np.random.RandomState(0)
+    picks = rng.permutation(np.arange(len(raws[0].ch_names)))[:10]
+    if test_preloading:
+        buffer_fname = op.join(tempdir, 'buffer')
+        raws.append(reader(preload=buffer_fname, **kwargs))
+        raws.append(reader(preload=True, **kwargs))
+
+        if test_blocks:
+            bnd = int(round(raws[0].info['buffer_size_sec'] *
+                            raws[0].info['sfreq']))
+        else:
+            bnd = raws[0].n_times // 2
+        slices = (slice(0, bnd), slice(bnd - 1, bnd), slice(3, bnd),
+                  slice(3, 300), slice(None))
+        if raws[0].n_times >= 2 * bnd:  # at least two complete blocks
+            slices = slices + (slice(bnd, 2 * bnd), slice(bnd, bnd + 1),
+                               slice(0, bnd + 100))
+        for sl_time in slices:
+            for raw in raws[1:]:
+                data1, times1 = raws[0][picks, sl_time]
+                data2, times2 = raw[picks, sl_time]
+                assert_allclose(data1, data2)
+                assert_allclose(times1, times2)
+
+    raw = raws[-1]  # use preloaded raw
+    full_data = raw._data
+
+    print(raw)  # to test repr
```

Comment:
We could do something a bit more clever like `assert_true(raw.__class__.__name__ in repr(raw))` to make sure we're constructing it properly. Same with `raw.info` repr check.